packages/protocol/lib/compatibility/ast-code.ts

```diff
@@ -0,0 +1,425 @@
+  const state2 = m2.stateMutability
+  if (state1 !== state2) {
+    report.push(new MethodMutabilityChange(contract, signature, state1, state2))
+  }
+  return report
+}
+
+const getCheckableMethodsFromAST = (contract: ContractAST, id: string): any[] => {
+  const checkableMethods = (method: Method) => method.visibility === VISIBILITY_EXTERNAL || method.visibility === VISIBILITY_PUBLIC
+  try {
+    return contract.getMethods().filter(checkableMethods)
+  } catch (error) {
+    throw {
+      message: `Error in the @openzeppelin/.../ContractAST.getMethods() for the artifacts in the '${id}' folder. 
+    Most likely this is due to a botched build, or a build on a non-cleaned folder.`,
+      error
+    }
+  }
+}
+
+function doASTCompatibilityReport(
+  contractName: string,
+  oldAST: ContractAST,
+  newAST: ContractAST): ASTCodeCompatibilityReport {
+  const oldMethods = createMethodIndex(getCheckableMethodsFromAST(oldAST, 'old'))
+  const newMethods = createMethodIndex(getCheckableMethodsFromAST(newAST, 'new'))
+
+  const report = new ASTCodeCompatibilityReport([])
+
+  // Check for modified or missing methods in the new version
+  Object.keys(oldMethods).forEach((signature: string) => {
+    const method: Method = oldMethods[signature]
+    if (!newMethods.hasOwnProperty(signature)) {
+      // Method deleted, major change
+      report.push(new MethodRemovedChange(contractName, signature))
+      // Continue
+      return
+    }
+    const newMethod: Method = newMethods[signature]
+    report.include(checkMethodCompatibility(contractName, method, newMethod))
+  })
+  // Check for added methods in the new version
+  Object.keys(newMethods).forEach((signature: string) => {
+    if (!oldMethods.hasOwnProperty(signature)) {
+      // New method, minor change
+      report.push(new MethodAddedChange(contractName, signature))
+    }
+  })
+  return report
+}
+
+function generateASTCompatibilityReport(oldContract: ZContract, oldArtifacts: BuildArtifacts,
+  newContract: ZContract, newArtifacts: BuildArtifacts): ASTCodeCompatibilityReport {
+  // Sanity checks
+  if (newContract === null) {
+    throw new Error('newContract cannot be null')
+  }
+  if (oldArtifacts === null) {
+    throw new Error('oldArtifacts cannot be null')
+  }
+  if (newArtifacts === null) {
+    throw new Error('newArtifacts cannot be null')
+  }
+  const contractName = newContract.schema.contractName
+
+  // Need to manually use ContractAST since its internal use in ZContract
+  // does not pass the artifacts parameter to the constructor, therefore
+  // forcing a reloading of BuildArtifacts.
+  const newAST = new ContractAST(newContract, newArtifacts)
+  const newKind = newAST.getContractNode().contractKind
+  if (oldContract === null) {
+    if (newKind === CONTRACT_KIND_CONTRACT) {
+      return new ASTCodeCompatibilityReport([new NewContractChange(contractName)])
+    } else {
+      // New contract added of a non-contract kind (library/interface)
```

Comment:
How does this tool handle changes to linked (rather than inlined) libraries?

Reply:
Great question, can you point me to an example of using a linked library?

Reply:
Ok so since linked libraries are only a name placeholder, if the linked library name doesn't change, the tool can't spot any change, since, the bytecode is exactly the same, and the change would only be a deployment change.

For our specific needs in cLabs, we may want to do some extra work here to manage versioning for contracts that depend on modified linked libraries.